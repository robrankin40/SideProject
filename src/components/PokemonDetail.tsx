import React from 'react';
import {Pokemon} from '../types';
import {View, Text, StyleSheet, Image} from 'react-native';

interface PokemonDetailProps {
  detail: Pokemon;
}

export const PokemonDetail: React.FC<PokemonDetailProps> = ({detail}) => {
  return (
    <View style={styles.wrapper}>
      <Text>{detail.name}</Text>
      <Text>Base Experience: {detail.base_experience}</Text>
      <Text>Weight: {detail.weight}</Text>
      <Text>Height: {detail.height}</Text>
      <Text>
        Abilities:{' '}
        {detail.abilities.map(({ability}) => ability.name).join(', ')}
      </Text>
      <Text>Forms: {detail.forms.map(({name}) => name).join(', ')}</Text>
      <Text>Moves: {detail.moves.map(({move}) => move.name).join(', ')}</Text>
      <Image
        source={{uri: detail.sprites.front_default}}
        style={styles.sprite}
      />
      <Image
        source={{uri: detail.sprites.back_default}}
        style={styles.sprite}
      />
    </View>
  );
};

const styles = StyleSheet.create({
  wrapper: {
    margin: 16,
    padding: 16,
    borderRadius: 8,
    borderWidth: 1,
    borderColor: 'black',
  },
  sprite: {
    width: 128,
    height: 128,
  },
});
