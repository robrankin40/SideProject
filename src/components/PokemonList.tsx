import React from 'react';
import {FlatList, StyleSheet, Text, TouchableOpacity, View} from 'react-native';

import {APIResult} from '../types';

export interface PokemonListProps {
  data?: APIResult[];
  loading: boolean;
  hasError: boolean;
  onPressPokemon?: (pokemon: string) => void;
}

const EmptyState: React.FC<{loading: boolean; hasError: boolean}> = ({
  loading,
  hasError,
}) => (
  <View>
    {loading ? (
      <Text>Loading...</Text>
    ) : hasError ? (
      <Text>Unexpected Error</Text>
    ) : (
      <Text>No Pokemons</Text>
    )}
  </View>
);

export const PokemonList: React.FC<PokemonListProps> = ({
  data,
  loading,
  hasError,
  onPressPokemon,
}) => {
  const renderItem = ({item}: {item: APIResult; index: number}) => (
    <TouchableOpacity
      style={styles.itemWrapper}
      onPress={() => onPressPokemon?.(item.name)}
      testID={`button_${item.name}`}>
      <Text>{item.name}</Text>
    </TouchableOpacity>
  );
  return (
    <FlatList
      data={data}
      renderItem={renderItem}
      keyExtractor={(item, index) => `pokemon_${item.name}_${index}`}
      ListEmptyComponent={<EmptyState loading={loading} hasError={hasError} />}
    />
  );
};

const styles = StyleSheet.create({
  itemWrapper: {
    padding: 16,
    margin: 16,
    borderRadius: 8,
    borderWidth: 1,
    borderColor: 'black',
  },
});
