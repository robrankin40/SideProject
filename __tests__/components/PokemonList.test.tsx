import React from 'react';
import {render, screen} from '@testing-library/react-native';
import {it, expect} from '@jest/globals';
import {mockPokemonList} from '../../mock';
import {PokemonList} from '../../src/components/PokemonList';

it('Pokemon List Renders Correctly with data', () => {
  render(
    <PokemonList
      data={mockPokemonList.results}
      onPressPokemon={(_: string) => {}}
      loading={false}
      hasError={false}
    />,
  );
  expect(screen).toMatchSnapshot();
});

it('Pokemon List Renders Correctly for loading', () => {
  render(
    <PokemonList
      data={[]}
      onPressPokemon={(_: string) => {}}
      loading
      hasError={false}
    />,
  );
  expect(screen).toMatchSnapshot();
});

it('Pokemon List Renders Correctly for error', () => {
  render(
    <PokemonList
      data={[]}
      onPressPokemon={(_: string) => {}}
      loading={false}
      hasError
    />,
  );
  expect(screen).toMatchSnapshot();
});
