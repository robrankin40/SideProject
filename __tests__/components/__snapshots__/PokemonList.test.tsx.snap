// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pokemon List Renders Correctly for error 1`] = `
<RCTScrollView
  ListEmptyComponent={
    <EmptyState
      hasError={true}
      loading={false}
    />
  }
  data={[]}
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={[]}
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View>
      <Text>
        Unexpected Error
      </Text>
    </View>
  </View>
</RCTScrollView>
`;

exports[`Pokemon List Renders Correctly for loading 1`] = `
<RCTScrollView
  ListEmptyComponent={
    <EmptyState
      hasError={false}
      loading={true}
    />
  }
  data={[]}
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={[]}
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View>
      <Text>
        Loading...
      </Text>
    </View>
  </View>
</RCTScrollView>
`;

exports[`Pokemon List Renders Correctly with data 1`] = `
<RCTScrollView
  ListEmptyComponent={
    <EmptyState
      hasError={false}
      loading={false}
    />
  }
  data={
    [
      {
        "name": "bulbasaur",
        "url": "https://pokeapi.co/api/v2/pokemon/1/",
      },
      {
        "name": "ivysaur",
        "url": "https://pokeapi.co/api/v2/pokemon/2/",
      },
      {
        "name": "venusaur",
        "url": "https://pokeapi.co/api/v2/pokemon/3/",
      },
      {
        "name": "charmander",
        "url": "https://pokeapi.co/api/v2/pokemon/4/",
      },
      {
        "name": "charmeleon",
        "url": "https://pokeapi.co/api/v2/pokemon/5/",
      },
      {
        "name": "charizard",
        "url": "https://pokeapi.co/api/v2/pokemon/6/",
      },
      {
        "name": "squirtle",
        "url": "https://pokeapi.co/api/v2/pokemon/7/",
      },
      {
        "name": "wartortle",
        "url": "https://pokeapi.co/api/v2/pokemon/8/",
      },
      {
        "name": "blastoise",
        "url": "https://pokeapi.co/api/v2/pokemon/9/",
      },
      {
        "name": "caterpie",
        "url": "https://pokeapi.co/api/v2/pokemon/10/",
      },
      {
        "name": "metapod",
        "url": "https://pokeapi.co/api/v2/pokemon/11/",
      },
      {
        "name": "butterfree",
        "url": "https://pokeapi.co/api/v2/pokemon/12/",
      },
      {
        "name": "weedle",
        "url": "https://pokeapi.co/api/v2/pokemon/13/",
      },
      {
        "name": "kakuna",
        "url": "https://pokeapi.co/api/v2/pokemon/14/",
      },
      {
        "name": "beedrill",
        "url": "https://pokeapi.co/api/v2/pokemon/15/",
      },
      {
        "name": "pidgey",
        "url": "https://pokeapi.co/api/v2/pokemon/16/",
      },
      {
        "name": "pidgeotto",
        "url": "https://pokeapi.co/api/v2/pokemon/17/",
      },
      {
        "name": "pidgeot",
        "url": "https://pokeapi.co/api/v2/pokemon/18/",
      },
      {
        "name": "rattata",
        "url": "https://pokeapi.co/api/v2/pokemon/19/",
      },
      {
        "name": "raticate",
        "url": "https://pokeapi.co/api/v2/pokemon/20/",
      },
    ]
  }
  getItem={[Function]}
  getItemCount={[Function]}
  keyExtractor={[Function]}
  onContentSizeChange={[Function]}
  onLayout={[Function]}
  onMomentumScrollBegin={[Function]}
  onMomentumScrollEnd={[Function]}
  onScroll={[Function]}
  onScrollBeginDrag={[Function]}
  onScrollEndDrag={[Function]}
  removeClippedSubviews={false}
  renderItem={[Function]}
  scrollEventThrottle={50}
  stickyHeaderIndices={[]}
  viewabilityConfigCallbackPairs={[]}
>
  <View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "borderColor": "black",
            "borderRadius": 8,
            "borderWidth": 1,
            "margin": 16,
            "opacity": 1,
            "padding": 16,
          }
        }
        testID="button_bulbasaur"
      >
        <Text>
          bulbasaur
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "borderColor": "black",
            "borderRadius": 8,
            "borderWidth": 1,
            "margin": 16,
            "opacity": 1,
            "padding": 16,
          }
        }
        testID="button_ivysaur"
      >
        <Text>
          ivysaur
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "borderColor": "black",
            "borderRadius": 8,
            "borderWidth": 1,
            "margin": 16,
            "opacity": 1,
            "padding": 16,
          }
        }
        testID="button_venusaur"
      >
        <Text>
          venusaur
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "borderColor": "black",
            "borderRadius": 8,
            "borderWidth": 1,
            "margin": 16,
            "opacity": 1,
            "padding": 16,
          }
        }
        testID="button_charmander"
      >
        <Text>
          charmander
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "borderColor": "black",
            "borderRadius": 8,
            "borderWidth": 1,
            "margin": 16,
            "opacity": 1,
            "padding": 16,
          }
        }
        testID="button_charmeleon"
      >
        <Text>
          charmeleon
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "borderColor": "black",
            "borderRadius": 8,
            "borderWidth": 1,
            "margin": 16,
            "opacity": 1,
            "padding": 16,
          }
        }
        testID="button_charizard"
      >
        <Text>
          charizard
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "borderColor": "black",
            "borderRadius": 8,
            "borderWidth": 1,
            "margin": 16,
            "opacity": 1,
            "padding": 16,
          }
        }
        testID="button_squirtle"
      >
        <Text>
          squirtle
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "borderColor": "black",
            "borderRadius": 8,
            "borderWidth": 1,
            "margin": 16,
            "opacity": 1,
            "padding": 16,
          }
        }
        testID="button_wartortle"
      >
        <Text>
          wartortle
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "borderColor": "black",
            "borderRadius": 8,
            "borderWidth": 1,
            "margin": 16,
            "opacity": 1,
            "padding": 16,
          }
        }
        testID="button_blastoise"
      >
        <Text>
          blastoise
        </Text>
      </View>
    </View>
    <View
      onFocusCapture={[Function]}
      onLayout={[Function]}
      style={null}
    >
      <View
        accessibilityState={
          {
            "busy": undefined,
            "checked": undefined,
            "disabled": undefined,
            "expanded": undefined,
            "selected": undefined,
          }
        }
        accessibilityValue={
          {
            "max": undefined,
            "min": undefined,
            "now": undefined,
            "text": undefined,
          }
        }
        accessible={true}
        collapsable={false}
        focusable={true}
        onClick={[Function]}
        onResponderGrant={[Function]}
        onResponderMove={[Function]}
        onResponderRelease={[Function]}
        onResponderTerminate={[Function]}
        onResponderTerminationRequest={[Function]}
        onStartShouldSetResponder={[Function]}
        style={
          {
            "borderColor": "black",
            "borderRadius": 8,
            "borderWidth": 1,
            "margin": 16,
            "opacity": 1,
            "padding": 16,
          }
        }
        testID="button_caterpie"
      >
        <Text>
          caterpie
        </Text>
      </View>
    </View>
    <View
      style={
        {
          "height": 0,
        }
      }
    />
  </View>
</RCTScrollView>
`;
